src/mlir/verona-mlir.cc

```diff
@@ -92,10 +109,113 @@ namespace
         outputFile = newName.c_str();
       }
     }
+  }
+
+  /// Check the value of command-line arguments, beyond the verification
+  /// provided by cl::opt.
+  ///
+  /// Prints an error message and returns a failure if an invalid combination is
+  /// used.
+  mlir::LogicalResult verifyCommandLine()
+  {
+    if (splitInputFile && inputKind != InputKind::MLIR)
+    {
+      llvm::errs() << "-split-input-file can only be used with MLIR inputs\n";
+      return mlir::failure();
+    }
+
+    if (verifyDiagnostics && inputKind != InputKind::MLIR)
+    {
+      llvm::errs() << "-verify-diagnostics can only be used with MLIR inputs\n";
+      return mlir::failure();
+    }
+
+    return mlir::success();
+  }
+
+  mlir::LogicalResult openInput(std::unique_ptr<llvm::MemoryBuffer>* input)
+  {
+    std::string err;
+    *input = mlir::openInputFile(inputFile, &err);
+    if (!*input)
+    {
+      llvm::errs() << err << "\n";
+      return mlir::failure();
+    }
+
+    return mlir::success();
+  }
+
+  mlir::LogicalResult openOutput(std::unique_ptr<llvm::ToolOutputFile>* output)
+  {
+    std::string err;
+    *output = mlir::openOutputFile(outputFile, &err);
+    if (!*output)
+    {
+      llvm::errs() << err << "\n";
+      return mlir::failure();
+    }
+
+    return mlir::success();
+  }
+
+  mlir::LogicalResult processVeronaInput(llvm::raw_ostream& output)
```

Comment:
This pattern now solidifies the clear separation of error handling, not only in different components, but based on different input arguments, which creates really confusing behaviour as well as making it impossible for any tooling using the API to have a sane error handling mechanism.